### PR TITLE
[DEV] Support multiple targets

### PR DESCRIPTION
**Pull request by Louson**
Enable the build of several targets in the same build directory. The field
'target' in the menu can be a single string or an array of string. If given an
array, all entries will be built (the build order is undefined).

This does not affect the previous 'target' behaviour.

### CODE REVIEW

**Comment by Louson** [comment]
Linked to issue #103

**Comment by Louson** [comment]
An undefined behavior remains concerning the inheritance:

### Inherit from a virtual build configuration an set a bitbake target

This removes the virtual bitbake target

### Inherit from a normal build configuration and set a bitbake target

This removes previous targets, but maybe we want to merge them ?

### Inherit from multiple virtual build configurations

Should we merge all bitbake targets ?

### Inherit from multiple build configurations with at least one non virtual

Should we merge all non virtual bitbake targets ?

File: cooker/cooker.py [hunk 1]
```diff
@@ -179,21 +179,27 @@ def __init__(self, name, config, layers, local_conf, target, inherit):
         self.config_ = config
         self.layers_ = layers
         self.local_conf_ = local_conf
-        self.target_ = target
+        if type(target) == list:
+            self.targets_ = target
+        elif target:
+            self.targets_ = []
+            self.targets_.append(target)
```
Comment (pboettch) [inline cooker/cooker.py:186]:
```suggestion
            self.targets_ = [target]
```

File: cooker/cooker.py [hunk 2]
```diff
@@ -179,21 +179,27 @@ def __init__(self, name, config, layers, local_conf, target, inherit):
...
 
-    def target(self):
+    def targets(self):
         if not self.buildable():
             return None
 
         for build in [self] + self.ancestors_[::-1]:
-            if build.target_:
-                return build.target_
+            if len(build.targets_) > 0:
```
Comment (pboettch) [inline cooker/cooker.py:201]: Pythonic way of checking list emptiness.

```suggestion
            if build.targets_:
```

File: cooker/cooker.py [hunk 3]
```diff
@@ -487,8 +493,8 @@ def show(self, builds, layers, conf, tree, build_arg, sources):
         for build_name in sorted(builds):
             build = BuildConfiguration.ALL[build_name]
 
-            if build.target():
-                build_info = ' (bakes {})'.format(build.target())
+            if build.targets():
+                build_info = ' (bakes {})'.format(build.targets())
```
Comment (pboettch) [inline cooker/cooker.py:497]: I'd prefer to not simply dump the list, but generate a comma-separated string. `', '.join(build.targets())`

**Review by pboettch — CHANGES_REQUESTED** [review]

**Comment by pboettch** [comment]
> An undefined behavior remains concerning the inheritance:
> ### Inherit from a virtual build configuration an set a bitbake target
> 
> This removes the virtual bitbake target
> ### Inherit from a normal build configuration and set a bitbake target
> 
> This removes previous targets, but maybe we want to merge them ?
> ### Inherit from multiple virtual build configurations
> 
> Should we merge all bitbake targets ?
> ### Inherit from multiple build configurations with at least one non virtual
> 
> Should we merge all non virtual bitbake targets ?

I think targets should be overridden by the final build. As you have done your implementation.

Parent build-targets have most like no sens in this inherited build-configuration. At least not for image-targets. 

What is actually your use-case? What are the multiple targets you are building within one build-config/build-dir using the same local.conf and thus machine.conf etc. I only can think of non-image-targets. Please enlighten us.

**Comment by Louson** [comment]
Well, there is nothing I want to do with inheritance. I was just wondering.

In my case I am creating training courses. I will use the beaglebone for all projects. I'd like to build them all together.

**Comment by pboettch** [comment]
After discussing with @cpb- we understood that currently the target-field is overwritten if a parent has it. Your patch does the same with list of targets. So it's fine, let's discuss this in another thread or issue if this needs a change.

**Review by pboettch — APPROVED** [review]

**Comment by Louson** [comment]
Yes, same machine.conf, same local.conf, same layers

**Comment by pboettch** [comment]
Just with different package selections, or something different?

**Pull request merged**